src/main/java/com/timgroup/statsd/CgroupReader.java

```diff
@@ -1,62 +1,108 @@
+ * cgroup file.
  *
  */
 class CgroupReader {
     private static final Path CGROUP_PATH = Paths.get("/proc/self/cgroup");
+    /**
+     * DEFAULT_CGROUP_MOUNT_PATH is the default cgroup mount path.
+     **/
+    private static final Path DEFAULT_CGROUP_MOUNT_PATH = Paths.get("/sys/fs/cgroup");
+    /**
+     * CGROUP_NS_PATH is the path to the cgroup namespace file.
+     **/
+    private static final Path CGROUP_NS_PATH = Paths.get("/proc/self/ns/cgroup");
     private static final String CONTAINER_SOURCE = "[0-9a-f]{64}";
     private static final String TASK_SOURCE = "[0-9a-f]{32}-\\d+";
     private static final Pattern LINE_RE = Pattern.compile("^\\d+:[^:]*:(.+)$", Pattern.MULTILINE | Pattern.UNIX_LINES);
-    private static final Pattern CONTAINER_RE =
-        Pattern.compile(
+    private static final Pattern CONTAINER_RE = Pattern.compile(
             "(" + CONTAINER_SOURCE + "|" + TASK_SOURCE + ")(?:.scope)?$");
 
+    /**
+     * CGROUPV1_BASE_CONTROLLER is the controller used to identify the container-id
+     * in cgroup v1 (memory).
+     **/
+    private static final String CGROUPV1_BASE_CONTROLLER = "memory";
+    /**
+     * CGROUPV2_BASE_CONTROLLER is the controller used to identify the container-id
+     * in cgroup v2.
+     **/
+    private static final String CGROUPV2_BASE_CONTROLLER = "";
+    /**
+     * HOST_CGROUP_NAMESPACE_INODE is the inode of the host cgroup namespace.
+     **/
+    private static final long HOST_CGROUP_NAMESPACE_INODE = 0xEFFFFFFBL;
+
     private boolean readOnce = false;
+    /**
+     * containerID holds either the container ID or the cgroup controller inode.
+     **/
     public String containerID;
 
     /**
-     * Parses /proc/self/cgroup and returns the container ID if available.
+     * Returns the container ID if available or the cgroup controller inode.
      * 
-     * @throws IOException
-     *     if /proc/self/cgroup is readable and still an I/O error occurs reading from the stream
+     * @throws IOException if /proc/self/cgroup is readable and still an I/O error
+     *                     occurs reading from the stream.
      */
     public String getContainerID() throws IOException {
         if (readOnce) {
             return containerID;
         }
 
-        containerID = read(CGROUP_PATH);
+        final String cgroupContent = read(CGROUP_PATH);
+        if (cgroupContent == null || cgroupContent.isEmpty()) {
+            return null;
+        }
+        containerID = parse(cgroupContent);
+        /*
+         * If the container ID is not available it means that the application is either
+         * not running in a container or running is private cgroup namespace, we
+         * fallback to the cgroup controller inode. The agent (7.51+) will use it to get
+         * the container ID.
+         * 
+         */
+        if ((containerID == null || containerID.equals("")) && !isHostCgroupNamespace(CGROUP_NS_PATH)) {
```

Comment:
Can you please add a comment explaining why we exclude `isHostCgroupNamespace(CGROUP_NS_PATH)` case?